src/derivative_utils.jl

```diff
@@ -70,6 +82,31 @@ either ForwardDiff or finite difference will be used depending on the
 `jac_config` of the cache.
 """
 function calc_J!(integrator, cache::OrdinaryDiffEqMutableCache, is_compos)
+  if isdefined(cache, :nlsolver)
+    calc_J!(cache.nlsolver, integrator, cache, is_compos)
+  elseif isdefined(cache, :J)
+    calc_J_in_cache!(integrator, cache, is_compos)
+  else
+    error("No J found in the cache")
+  end
+end
+
+function calc_J!(nlsolver::NLSolver, integrator, cache::OrdinaryDiffEqMutableCache, is_compos)
```

Comment:
why not just pass J and have 1 function?

Reply:
That's because we unpack uf,du1,jac_config inside these functions, passing them in function too will make a lot of arguments.

Reply:
Speaking of arguments, is there any reason for passing around `is_compos`? Its value should only depend on `integrator`, so it seems a bit redundant?